File: monitor.go
package phi

import (
	"math"
	"time"
)

// TODO: Mean Squared Error
type monitor struct {
	*linear
	sum, sos float64 // sum of square
}

func newMonitor(size int, lf, tf float64) *monitor {
	return &monitor{linear: newLinear(size, lf, tf)}
}

func (m *monitor) Put(d time.Duration) {
	x := m.put(milliSeconds(d))
	m.sum += x
	m.sos += x * x
}

func (m *monitor) Phi() Value {
	i := len(m.signals) - 1
	x := m.level(i) + m.trend(i)
	return m.estimate(x)
}

func (m *monitor) estimate(x float64) Value {
	var (
		length   = float64(len(m.levels))
		mean     = m.sum / length
		variance = m.sos/length - mean*mean
		stddev   = math.Sqrt(variance)
	)
	return Value(calculate(x, mean, stddev))
}

// The value of phi is calculated as:
//
//    -math.Log10(1 - CDF(timeSinceLastHeartbeat)
//
// where CDF is the cumulative distribution function of a normal distribution
// with mean and stddev estimated from historical heartbeat inter-arrival durations
func calculate(x, mean, stddev float64) (val float64) {
	y := (x - mean) / stddev
	e := math.Exp(-y * (1.5976 + 0.070566*y*y))
	if x > mean {
		val = math.Max(0, -math.Log10(e/(1.0+e)))
	} else {
		val = math.Max(0, -math.Log10(1.0-1.0/(1.0+e)))
	}
	return
}

func (x *monitor) shrink(to int) {
	n := shrink(&x.signals, to)
	if n == 0 {
		return
	}
	x.sum, x.sos = 0, 0
	if to == 0 {
		x.levels, x.trends = nil, nil
		return
	}
	for i := 0; i < len(x.signals); i++ {
		x.putn(i, x.signals[i])
	}
}

File: src/monitor.rs
// Monitor implements `Holt's Linear Method` (DoubleExponentialSmoothing).
// `Holt's Linear Method` is good for non-seasonal data with a trend.
//
//    Level:    L[i]   = a*X[i] + (1−a)*(L[i-1] + T[i-1])
//    Trend:    T[i]   = b*(L[i] − L[i−1]) + (1−b)*T[i−1]
//    Forecast: F[i+1] = L[i] + T[i]
//
// a and b are factor for smoothing observed signals.
pub struct Monitor {
    cap: usize,
    sum: f64,
    sos: f64, // sum of square

    signals: Vec<f64>,
    lf: f64,
    tf: f64,
    levels: Vec<f64>,
    trends: Vec<f64>,
}

impl Monitor {
    pub fn new(cap: usize, lf: f64, tf: f64) -> Self {
        let signals = Vec::with_capacity(cap);
        let sum = 0.0;
        let sos = 0.0;
        let lf = bounded(lf);
        let tf = bounded(tf);
        let levels = Vec::new();
        let trends = Vec::new();
        Monitor {
            cap,
            sum,
            sos,
            signals,
            lf,
            tf,
            levels,
            trends,
        }
    }

    fn level(&self, i: usize, x: f64) -> f64 {
        if i == 0 {
            x
        } else {
            self.lf * x + (1. - self.lf) * (self.levels[i - 1] + self.trends[i - 1])
        }
    }

    fn trend(&self, i: usize, x: f64) -> f64 {
        if i == 0 {
            0.
        } else {
            self.tf * (x - self.levels[i - 1]) + (1. - self.tf) * self.trends[i - 1]
        }
    }

    pub fn push(&mut self, x: f64) {
        self.signals.push(x);
        let i = self.signals.len() - 1;
        let x = self.pushn(i, x);
        self.sum += x;
        self.sos += x * x;
    }

    fn pushn(&mut self, i: usize, x: f64) -> f64 {
        let l1 = self.level(i, x);
        self.levels.push(l1);
        let l2 = self.levels[i];
        let t = self.trend(i, l2);
        self.trends.push(t);
        l2
    }

    pub fn estimate(&self, d: f64) -> f64 {
        let len = self.levels.len() as f64;
        let mean = self.sum / len;
        let var = self.sos / len - mean * mean;
        let stddev = var.sqrt();
        phi_value(mean, stddev, d)
    }

    fn failure(&self) -> f64 {
        let i = self.signals.len();
        let x = self.levels[i] + self.trends[i];
        unimplemented!()
    }
}

fn phi_value(mean: f64, stddev: f64, d: f64) -> f64 {
    let y = (d - mean) / stddev;
    let p = (-y * (1.5976 + 0.070566 * y * y)).exp();
    if d > mean {
        -(p / (1.0 + p)).log10()
    } else {
        -(1.0 - 1.0 / (1.0 + p)).log10()
    }
}

fn bounded(f: f64) -> f64 {
    if f >= 1.0 {
        0.999
    } else if f <= 0. {
        0.100
    } else {
        f
    }
}
